website/content/2025-10-20-claude-code-plugin.md

```diff
@@ -0,0 +1,104 @@
+
+1. Explore the provided code and identify candidate properties.
+2. Explore how the codebases calls that code in practice.
+3. Grounded in this understanding, write corresponding Hypothesis tests.
+4. Run the new Hypothesis tests, and reflect on any failures. Is it a genuine bug, or is the test incorrect? Refactor the test if necessary.
+
+The legwork that `/hypothesis` instructs the model to do both before and after writing a test is critical for deriving high-quality tests. For example, the model might discover in step 2 that a function is called with two different input formats, and both should be tested. Or it might discover in step 4 that it wrote an unsound test, by generating test inputs the function didn't expect, like `math.nan`.
+
+## Failure modes
+
+We observed a few failure modes while developing `/hypothesis`. For example, AI models like to write strategies with unnecessary restrictions, like limiting the maximum length of a list even when the property should hold for all lengths of lists. We added explicit instructions in `/hypothesis` not to do this, though that doesn't appear to have fixed the problem entirely.
+
+By far the most fundamental failure mode is that the model might simply misunderstand a property in the code. For example, we ran `/hypothesis` on [python-dateutil](https://github.com/dateutil/dateutil); specifically, `/hypothesis src/easter.py`. The model determined that a property of the `easter` function is that it should always return a date on a Sunday, no matter the `method` argument, of which dateutil provides three: `method=EASTER_JULIAN`, `method=EASTER_ORTHODOX`, `method=EASTER_WESTERN`. The model wrote a test saying as much, which then failed, and it proudly claimed it had found a bug.
+
+In fact, the model had not found a bug. In reality, `dateutil.easter` computes the date for Easter in the calendar corresponding to the passed `method`, but always returns that date in the Gregorian calendar—which might not be a Sunday. The test written by the model assumed the computation occurred in the Gregorian calendar from start to finish, which was incorrect.
+
+This kind of subtle semantic reasoning remains difficult for models, and it's important to keep it in mind as a limitation.
+
+# Using `/hypothesis` for bug hunting
+
+As mentioned before, we [wrote a paper](https://mmaaz-git.github.io/agentic-pbt-site/) based on the `/hypothesis` command, where we use `/hypothesis` with Claude Sonnet 4 to write and run Hypothesis tests for popular Python packages. We found bugs in NumPy, pandas, and Google and Amazon SDKs, and [submitted](https://github.com/numpy/numpy/pull/29609) [patches](https://github.com/aws-powertools/powertools-lambda-python/pull/7246) [for](https://github.com/aws-cloudformation/cloudformation-cli/pull/1106) [a](https://github.com/huggingface/tokenizers/pull/1853) number of them. The paper is quite short, so do give it a read if you're interested.
```

Comment:
Opus 4.1 and Sonnet 4.5 (not Sonnet 4); and "several" of them lands the links more pleasingly imo.